### ISSUE TITLE
[BUG] Exaggerated ranges after optimizing

### ISSUE THREAD

**Issue by JeffersonCorreia32a** [open-spaced-repetition/fsrs-optimizer#5]
**Research**

*Enter an [x] character to confirm the points below:*

- [x ] I have checked the [FAQ](https://github.com/open-spaced-repetition/fsrs4anki/wiki/FAQ) and could not find an answer to my question
- [x ] I have read the [wiki](https://github.com/open-spaced-repetition/fsrs4anki/wiki) and still felt confused
- [x ] I have searched for similar existing questions here

**Question**

Dear all,

I would like to understand why my deck is displaying such absurd values. The optimizer is providing parameters that deviate significantly from the original values. When I optimize the deck, the "w" values increase to an unreasonable extent, causing the application of these values to result in absurdly long intervals between card reviews. I don't understand the underlying problem.

The parameters I used were as follows:

Next Day Starts at (Step 2): 12
Timezone (Step 3.1): America/Bahia
Filter out suspended cards: checked
Advanced Settings (Step 3.2): 0.9
Revlog Start Date: Optimize review logs after this date: 2006-10-05
Example original "w" values: If I keep the original "w" as follows:
[0.4, 0.6, 2.4, 5.8, 4.93, 0.94, 0.86, 0.01, 1.49, 0.14, 0.94, 2.18, 0.05, 0.34, 1.26, 0.29, 2.61]
A new card will have a good interval of 2 days and an easy interval of 7 days.

Example optimized "w" values: If I use the optimized "w" for my deck as follows:
[1.46, 4.9, 8.82, 10.08, 4.8994, 0.8604, 0.8373, 0.007, 1.5252, 0.1, 0.9738, 2.2595, 0.0204, 0.3773, 1.5048, 0.3335, 2.3037]
A new card will have a good interval of 10 days and an easy interval of 12 days.

A good interval of 10 days for a new card seems clearly absurd...

I would like to know if others are experiencing such significant deviations from the original parameters or if it's just affecting my deck. This might help identify the issue.

This problem has been occurring for me since version 3, but it seems to have worsened recently.

I believe it would be helpful to find a way to maintain these parameters within an acceptable standard deviation for a human. In my opinion, it's unrealistic for a human to fully memorize content from today to 10 days in the future after seeing it only once. Perhaps, if the optimizer could set a maximum good interval of 3 or 4 days within a standard deviation, it would be more reasonable.

These discrepancies persist in older reviews, resulting in significantly increased intervals. Additionally, there is a substantial difference between the "good" and "easy" buttons. For example, one card had a good interval of 29 days while the easy interval was 2.6 months, representing a 250% difference.

I have included my complete deck, including media, for your evaluation. Thank you in advance for any effort made to help solve this case.

[0PROJETO CONCURSO 20mil+26-07-2023 com midia.zip](https://github.com/open-spaced-repetition/fsrs4anki/files/12171189/0PROJETO.CONCURSO.20mil%2B26-07-2023.com.midia.zip)

**Comment by L-M-Sherlock** [comment]
I replaced `curve_fit` with `minimize`. It is more flexible and customizable.

https://github.com/open-spaced-repetition/fsrs-optimizer/blob/723da1e03a00bfa336410a17ba9f1dc931db070f/src/fsrs_optimizer/fsrs_optimizer.py#L510-L524

In this implementation, I set the `x0` and `maxiter`. It is conservative when the number of reviews is small. We could also add some penalty for large stability in the loss function.

**Comment by AkiraChisaka** [comment]
@L-M-Sherlock
> Could you set `verbose=True` for `optimizer.pretrain(verbose=False)` in the section 2.1 when you run the optimizer? We can check your forgetting curve for first easy rating.

Ok, I gave it a go, let me see... I'm not super familiar with Google Colabs yet, so I'll just paste the stuff in section 2.1 here:

---

```
optimizer.define_model()
optimizer.pretrain(verbose=True)
optimizer.train(verbose=False)
```
![image](https://github.com/open-spaced-repetition/fsrs-optimizer/assets/47231500/168813b5-b6d8-4e33-a7ca-480f4d8e2f6b)
{1: 0.128602622444628}
Not enough data for first rating 2. Expected at least 100, got 8.
![image](https://github.com/open-spaced-repetition/fsrs-optimizer/assets/47231500/82913a9e-38eb-460a-97b6-9173cf054682)
{1: 0.128602622444628, 3: 2.2090953146625885}
Not enough data for first rating 4. Expected at least 100, got 83.
Weighted fit parameters: [-10.08928228   3.6075409    0.12707148]
Fit stability: [0.12860262 2.20909531]
RMSE: 0.0000
![image](https://github.com/open-spaced-repetition/fsrs-optimizer/assets/47231500/512d5eab-0948-4a33-a9dc-fa8b5fb317ae)

---

Yeah... this might be starting to look a bit buggy. Since I think rating 4 means "easy", and I am `Not enough data for first rating 4. Expected at least 100, got 83.`.

And the optimizer kinda just decided "83/100? Yeah that's good enough let's give it a solid 30.0 out of 30.0"

**Comment by user1823** [comment]
> I replaced `curve_fit` with `minimize`. It is more flexible and customizable.

I tested this (by copying the relevant portions into fsrs4anki_optimizer_beta.ipynb) and also tried adjusting the values of `maxiter`. My results are tabulated below.

Setting | Initial Stability
-- | --
Current optimizer | {1: 1.17, 2: 6.93, 3: 16.65, 4: 30.0}
maxiter = total_count | {1: 1.17, 2: 10.85, 3: 16.64, 4: 160.65}
maxiter = total_count / 10 | {1: 1.17, 2: 1.19, 3: 16.64, 4: 160.65}
maxiter = total_count / 15 | {1: 1.17, 2: 1.04, 3: 16.64, 4: 69.91}
maxiter = total_count / 20 | {1: 1.17, 2: 1.04, 3: 16.64, 4: 39.76}
maxiter = total_count / 50 | {1: 1.17, 2: 1.04, 3: 16.64, 4: 13.84}
maxiter = total_count / 100 | {1: 1.17, 2: 1.04, 3: 16.64, 4: 8.73}

For my collection, I find the results of `maxiter = total_count / 15` and `maxiter = total_count / 20` to be more or less acceptable. But we will have to find a solution that works well for all the users.

**Comment by Expertium** [comment]
Try `maxiter = int(np.sqrt(total_count))`, I'm curious.

**Comment by L-M-Sherlock** [comment]
I have added L1 regularization for initial stability. It performs better.

**Comment by L-M-Sherlock** [comment]
```
%pip install git+https://github.com/open-spaced-repetition/fsrs-optimizer@Fix/exaggerated-initial-stability-after-optimizing
```

The line of code could install the package in the notebook.

**Comment by Expertium** [comment]
@JeffersonCorreia32a Sherlock changed the way S0 is calculated, try it again.

**Comment by user1823** [comment]
> I have added L1 regularization for initial stability. It performs better.

L1 regularization increased the RMSE for me.

Loss = MSE
```
{1: 1.17, 2: 2.47, 3: 16.64, 4: 83.19}

R-squared: 0.9487
RMSE: 0.0113

Last rating: 1
R-squared: 0.7937
RMSE: 0.0267

Last rating: 2
R-squared: -0.2459
RMSE: 0.1157

Last rating: 3
R-squared: 0.9622
RMSE: 0.0091

Last rating: 4
R-squared: -27.4365
RMSE: 0.0397
```

Loss = MSE + L1
```
{1: 1.17, 2: 1.74, 3: 8.49, 4: 16.6}

R-squared: 0.9443
RMSE: 0.0114

Last rating: 1
R-squared: 0.8448
RMSE: 0.0227

Last rating: 2
R-squared: -0.3572
RMSE: 0.1213

Last rating: 3
R-squared: 0.9345
RMSE: 0.0118

Last rating: 4
R-squared: -246.6033
RMSE: 0.0923
```

All the other factors were the same i.e. s0 for Hard rating = 1.5 and `maxiter = int(np.sqrt(total_count)`.

**Comment by Expertium** [comment]
I would do proper tests of statistical significance with many decks and collections. but it's too much of a pain without having a version of the optimizer with visible code.

**Comment by L-M-Sherlock** [comment]
Could you use the command line version to test statistical significance? It's faster and you can also modify the code.

```shell
python ./src/fsrs_optimizer/__main__.py "./"
```

This command could optimizer all decks and collections in the current directory.

**Comment by Expertium** [comment]
I have a better idea. Here's the code:
```
import scipy
list1 = []
list2 = []
wilcox = scipy.stats.wilcoxon(array1, array2).pvalue
```
list1 and list2 should:
1) have the same length
2) contain values ordered in the same way. So if list1 has a1, a2, a3...an, then in list2 the values should also be b1, b2, b3...bn.
One of the lists should contain baseline values of RMSE, and the other should contain values of RMSE after making a change. If a1 is the RMSE of some deck, then b1 should be the RMSE **of the same deck after** making a change to the optimizer. Not that a low p-value doesn't tell you which one has lower average RMSE, that's something you'll have to calculate separately to find out which one is better.

**Comment by L-M-Sherlock** [comment]
I did the tests between 4.5.3 and current branch:

|                                              | old RMSE   | new  RMSE  | old init S                     | new init S                     |
| -------------------------------------------- | ------ | ------ | ------------------------- | ------------------------- |
| collection-2022-09-18@13-21-58               | 0.0154 | 0.0157 | 1.14, 1.01, 5.44, 14.67   | 1.14, 1.01, 5.43, 14.11   |
| Main_27.04.2023                              | 0.0549 | 0.0544 | 0.1, 0.32, 1.14, 196.37   | 0.6, 0.32, 1.14, 147.46   |
| Default                                      | 0.0128 | 0.0128 | 1.0, 2.0, 6.81, 30.0      | 1.0, 1.46, 6.81, 34.62    |
| 0PROJETO CONCURSO 20mil+26-07-2023 com midia | 0.0355 | 0.0345 | 1.46, 4.9, 8.82, 10.08    | 1.45, 2.75, 7.84, 9.27    |
| English                                      | 0.0201 | 0.0182 | 23.47, 30.0, 22.37, 23.47 | 9.27, 32.73, 21.92, 25.33 |

p = 0.14

The improvement is not significant statistically. But it really decrease the deviation when the review counts is small.

**Comment by user1823** [comment]
> I did the tests between 4.5.3 and current branch:

What if you make the following changes?
- use `maxiter = int(np.sqrt(total_count)` and
- increase the S0 for Hard to 1.5 (0.6 is too small)

**Comment by L-M-Sherlock** [comment]
> > I did the tests between 4.5.3 and current branch:
> 
> What if you make the following changes?
> 
> * use `maxiter = int(np.sqrt(total_count)` and
> * increase the S0 for Hard to 1.5 (0.6 is too small)

|                                              | old    | new    | user1823 | old                       | new                       | user1823                 |
| -------------------------------------------- | ------ | ------ | -------- | ------------------------- | ------------------------- | ------------------------ |
| collection-2022-09-18@13-21-58               | 0.0154 | 0.0157 | 0.0157   | 1.14, 1.01, 5.44, 14.67   | 1.14, 1.01, 5.43, 14.11   | 1.14, 1.01, 5.43, 14.11  |
| Main_27.04.2023                              | 0.0549 | 0.0544 | 0.0544   | 0.1, 0.32, 1.14, 196.37   | 0.6, 0.32, 1.14, 147.46   | 0.6, 0.32, 1.14, 147.46  |
| Default                                      | 0.0128 | 0.0128 | 0.0128   | 1.0, 2.0, 6.81, 30.0      | 1.0, 1.46, 6.81, 34.62    | 1.0, 1.46, 6.81, 34.62   |
| 0PROJETO CONCURSO 20mil+26-07-2023 com midia | 0.0355 | 0.0345 | 0.0345   | 1.46, 4.9, 8.82, 10.08    | 1.45, 2.75, 7.84, 9.27    | 1.45, 2.4, 7.84, 9.27    |
| English                                      | 0.0201 | 0.0182 | 0.0223   | 23.47, 30.0, 22.37, 23.47 | 9.27, 32.73, 21.92, 25.33 | 4.1, 32.77, 21.92, 20.89 |

**Comment by user1823** [comment]
So, I guess that the effect which was to be achieved by using sqrt of total_count in maxiter has already been achieved by using L1 regularization. I think so because using sqrt in maxiter significantly affected the results in my testing yesterday.

But, I still suggest using `S0 = 1.5` for Hard.

Also, I am curious why the English deck has such high initial stability values.

**Comment by L-M-Sherlock** [comment]
> But, I still suggest using `S0 = 1.5` for Hard.

I have used it. But it seems not to affect the result.

**Comment by L-M-Sherlock** [comment]
> Also, I am curious why the English deck has such high initial stability values.

Here are some figures about the English deck:

![pretrain_1](https://github.com/open-spaced-repetition/fsrs-optimizer/assets/32575846/74544ee9-c5ed-48e2-abeb-82880f8c68bf)
![pretrain_2](https://github.com/open-spaced-repetition/fsrs-optimizer/assets/32575846/4db7981e-73cc-4cc0-b5b7-c4a4f08dbbe4)

I don't know why @AnyonicFugue has so high retention in some delta_ts.

**Comment by L-M-Sherlock** [comment]
> So I wonder is there a way to produce reasonable fsrs parameters. Should I delete the data points with exceptionally low retention and run the optimizer again?
> 
> Additional information that might be useful: I'm mainly using Anki to memorize English words to prepare for the GRE exam. The difficulty of different words can be very different for me. My memory is, I think, a bit above average, but far below the V4 optimizer predicts.

IMO, you may delete the data points with exceptionally **high** retention. And you should rate the cards according to their difficulty for you, especially in the first learning.

**Comment by user1823** [comment]
> > But, I still suggest using `S0 = 1.5` for Hard.
> I have used it. But it seems not to affect the result.

It might not affect our results, but it is a better default. I am sure that no one would be negatively affected by using `S0 = 1.5` for Hard (instead of 0.6) but some people might be positively affected.

> I don't know why @AnyonicFugue has so high retention in some delta_ts.

If you look carefully, the delta_t with higher retention are also those with low count. So, we definitely need to find a good way to deal with delta_t having low counts. It seems that additive smoothing is not working _that_ well.

To do this, the first step would be to analyse the stability_for_pretrain.tsv file.

For my collection, have a look at the following result:

![Screenshot](https://github.com/open-spaced-repetition/fsrs-optimizer/assets/92206575/74b2b27a-7ac6-4d84-809d-6ab29335f3fa)

Here, some delta_t have a retention of 0. Additive smoothing would significantly increase the retention and would, thus, lead to an overestimation of the initial stability.

**Comment by L-M-Sherlock** [comment]
```
{
    // Generated, Optimized anki deck settings
    "deckName": "collection-2023-07-29@16-31-40",// PLEASE CHANGE THIS TO THE DECKS PROPER NAME
    "w": [0.13, 0.6, 2.24, 6.31, 5.0146, 0.9854, 0.8846, 0.0, 1.3681, 0.1098, 0.8124, 2.219, 0.0275, 0.3536, 1.2753, 0.3298, 2.4546],
    "requestRetention": 0.8921632653061224,
    "maximumInterval": 36500,
},
```

The current version has solved the problem from @AkiraChisaka. The initial stability of `easy` is more rational.

**Comment by L-M-Sherlock** [comment]
> If you look carefully, the delta_t with higher retention are also those with low count. So, we definitely need to find a good way to deal with delta_t having low counts. It seems that additive smoothing is not working _that_ well.

I figure it out. The current version uses the average recall from entire collection. The average recall from the first reviews will be rational.

**Comment by L-M-Sherlock** [comment]
Oops. I'm wrong. His dataset is weirder than my thought.

<img width="262" alt="image" src="https://github.com/open-spaced-repetition/fsrs-optimizer/assets/32575846/9a2469fc-942e-4daf-92be-6f085784d8d4">

The `Additive smoothing` is not the main culprit.

**Comment by L-M-Sherlock** [comment]
I recommend merging this branch and opening a new issue to discuss exceptional case of @AnyonicFugue.

**Comment by Expertium** [comment]
@L-M-Sherlock do you only have 5 collections? Because that is a very small sample size, in order to calculate the p-value accurately more data is needed. I will share a few more collections later, if you have more collections - please use more.

**Comment by L-M-Sherlock** [comment]
I have 60+ deck/collections. But some of them are duplicated or overlapped. The current comparison need too many manual operation. I plan to automate the comparison.

**Comment by Expertium** [comment]
Here is some more data, though you probably already have most of these collections
https://drive.google.com/file/d/1veZtb3SlGxzQs5E_y5-5K22ZhEUikmdC/view?usp=sharing

**Comment by Expertium** [comment]
> The current version uses the average recall from entire collection. The average recall from the first reviews will be rational.

Also, this is a good idea, I think this should be implemented.
